### PR TITLE
Added support for symbolic links in import resolver

### PR DESCRIPTION
**Pull request by erictraut**
Added support for symbolic links in import resolver both for resolution of ".pth" files and for imports themselves.

### CODE REVIEW

File: packages/pyright-internal/src/analyzer/pythonPathUtils.ts [hunk 1]
```diff
@@ -268,26 +268,37 @@ function getPathsFromPthFiles(fs: FileSystem, parentDir: string): string[] {
...
-                if (trimmedLine.length > 0 && !trimmedLine.startsWith('#') && !trimmedLine.match(/^import\s/)) {
-                    const pthPath = combinePaths(parentDir, trimmedLine);
-                    if (fs.existsSync(pthPath) && isDirectory(fs, pthPath)) {
-                        searchPaths.push(pthPath);
+        let filePath = combinePaths(parentDir, pthFile.name);
+
+        if (pthFile.isSymbolicLink()) {
+            // Resolve the symlink.
+            filePath = fs.realpathSync(filePath);
+            if (!fs.existsSync(filePath)) {
```
[the remark this answers is not in the document]
Comment (jakebailey) [reply packages/pyright-internal/src/analyzer/pythonPathUtils.ts:280]: Ah, you check it afterward.

**Comment by jakebailey** [comment]
Some of this surprises me... `fs.statSync` should already be following symlinks, it's `lstat` call that explicitly doesn't follow them. I can understand it being required for places where we iterate over a directory (I don't think the entries you get from `readdir` have been stat'd), but in the other uses `stat` should never be able to return a symlink.

Also, I think it's best to always return the original path (don't resolve) to ensure later uses like filename to module name mappings work as intended, but it's also true that we can end up with aliasing problems now due to paths giving the same filename multiple ways to be imported.

**Comment by erictraut** [comment]
Are you sure that fs.statSync follows symlinks? Why does its return type ("Stats") have a method called "isSymbolicLink"? If it resolves links, this will always be false.

**Comment by jakebailey** [comment]
"Stat" is a single structure common to all stat calls, it's just that the `lstat` call stops following when it encounters a symlink, while `stat` always follows and never returns a symlink. If you look in the VFS, you'll see that they are the same underlying method, just with a "follow" boolean changed. That is of course the VFS we copied from TS (so isn't evidence), but this is documented in the actual syscall definitions themselves in POSIX which the NodeJS docs reference:

https://www.man7.org/linux/man-pages/man2/stat.2.html#DESCRIPTION
https://nodejs.org/api/fs.html#fs_fspromises_stat_path_options
https://nodejs.org/api/fs.html#fs_fspromises_lstat_path_options
https://stackoverflow.com/a/2635948

**Comment by jakebailey** [comment]
We didn't implement fourslash tests for soft/hard links (just regular files), but feasibly could to verify our behavior here. The VFS has the support for it (see the `Symlink` type).

**Comment by erictraut** [comment]
OK, then I significantly overcomplicated this problem. :) I'll abandon this PR and post a new one.

**Pull request closed without merging**